### ISSUE TITLE
Recommended way to work with torchaudio and torchaudio_contrib?

### ISSUE THREAD

**Issue by rfalcon100** [audeering/audtorch#46]
I am looking to use and expand some transforms for audio. However, what is the recommended way of working with audeering/audtorch [audtorch](https://github.com/audeering/audtorch),  [torchaudio](https://github.com/pytorch/audio) , and [torchaudio_contrib](https://github.com/keunwoochoi/torchaudio-contrib) ? 

1) Mostly, there are some functions that do the same thing (for example, stft, melspectrograms). So what the best way to use functions from different libraries? Is it ok to mix them?

2) When I write my I own transforms, should I clone this repo and add my code to there?

**Comment by hagenw** [comment]
It's indeed a little bit confusing with pytorch and audio. I would say in the long run you will most probably use [torchaudio](https://github.com/pytorch/audio), but at the moment its API is not fixed and it's unstable to work with.
[torchaudio_contrib](https://github.com/keunwoochoi/torchaudio-contrib) its not meant to be used directly, but to discuss the API of [torchaudio](https://github.com/pytorch/audio) and prepare contributions to it. [audtorch](https://github.com/audeering/audtorch) started due to the unstable status and lack of functionality of [torchaudio](https://github.com/pytorch/audio) with the goal of providing easy installation and a few transforms and data sets to start with.

In summary:
* If you would like to work with audio data sets like `LibriSpeech`, `VoxCeleb1`, or `MozillaCommonVoice` and if it's fine to  include `numpy` based packages in your project I would propose to use [audtorch](https://github.com/audeering/audtorch)
* If you need to load your audio directly as `torch.Tensor` and would like to integrate your transforms as layers into your model use [torchaudio](https://github.com/pytorch/audio)
* It's also totally fine to mix all of the available projects

Feel also free to discuss here in more detail what kind of transforms you would like to add, maybe it's easier then to answer where they would fit the best.

Regarding creating pull requests for adding code to [audtorch](https://github.com/audeering/audtorch), the proposed way to do it is to follow the [Creating a pull request from a fork](https://help.github.com/en/articles/creating-a-pull-request-from-a-fork) workflow.

If you plan to contribute a pull request to [torchaudio](https://github.com/pytorch/audio) you should maybe first check if similar stuff is already implemented in [torchaudio_contrib](https://github.com/keunwoochoi/torchaudio-contrib) and if yes ask them how you could help integrating it into [torchaudio](https://github.com/pytorch/audio).

**Comment by rfalcon100** [comment]
Thanks for the response. I understand the whole picture a bit better now. That said:

- What are the main differences between using `numpy` and `torch.Tensor` for loading the audio data? I have tried using the [torchaudio](https://github.com/pytorch/audio) library to load my own datasets, but loading and decoding audio files is quite slow. So is it better to do a preprocessing step to save all the audio data as either `numpy`or `torch.Tensor`?

- Is there an advantage on integrating transforms as layers? I am mostly familiar with the `torchvision` package where transforms as used only on the dataset and not on the model.

**Comment by phtephanx** [comment]
```
What are the main differences between using numpy and torch.Tensor for loading the audio data? I have tried using the torchaudio library to load my own datasets, but loading and decoding audio files is quite slow. So is it better to do a preprocessing step to save all the audio data as either numpyor torch.Tensor?
```
Are you referring to only loading audio files, e.g. using `torchaudio.load`. In our experience loading and additional preprocessing steps, e.g. stft, should certainly be fast enough to do it lazily in `__getitem__`.

```
Is there an advantage on integrating transforms as layers? I am mostly familiar with the torchvision package where transforms as used only on the dataset and not on the model.
```
When using transforms as layers, transform ops ...
* can be moved to GPU resources which are supposed to be scalable
* work on batches which can potentially be more efficient
* can be stored easily as part of the model for deployment

**Comment by rfalcon100** [comment]
> Are you referring to only loading audio files, e.g. using torchaudio.load. In our experience loading and additional preprocessing steps, e.g. stft, should certainly be fast enough to do it lazily in __getitem__.

Yes, I am having issues reading mp3 files. It appears that just loading the files is very slow, possibly due to the decoding process. So I will most likely preprocess all files to either `numpy` or `torch.Tesnor` files.

>When using transforms as layers, transform ops ...

Do you know a good reference on how to use transforms as layers? All the examples I have seem so far use transforms in the dataset and not the model.

**Comment by phtephanx** [comment]
```
Do you know a good reference on how to use transforms as layers? All the examples I have seem so far use transforms in the dataset and not the model.
```
Well, the transforms need to be designed as layers such as in:
https://github.com/keunwoochoi/torchaudio-contrib/blob/master/torchaudio_contrib/layers.py
Then, you can just use them like any other layer in your model.

Another way would be to use the "classic" transforms, make it a layer by dynamic inheritance. Something like that:
```
import numpy as np
import torch.nn as nn
from audtorch.transforms import Crop


def make_module(cls):
    class ModuleChild(nn.Module, cls):
        def __init__(self, *args, **kwargs):
            nn.Module.__init__(self)
            cls.__init__(self, *args, **kwargs)

        def forward(self, signal):
            return cls.__call__(self, signal)

        def __repr__(self):
            return cls.__repr__(self)

    return ModuleChild


t = make_module(Crop)(idx=0, axis=-1)
# t can be used as a layer now
a = np.array([[1, 2], [3, 4]])
print(t)
print(t(a))
```
Now one could use it as a layer like above. But you need to be very careful with respect to how the Callable works on the dimensions since the incoming tensor has now 1 additional dimension, i.e. the batch dimension. E.g. the `axis` parameter could take a different value.

What happens when using "classic" transforms as layers is that you apply the same operation (e.g. crop) on all samples of the batch. This might be desired or not. One would e.g. crop the samples of a batched tensor exactly the same.
